dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/DefaultTrackerBundleService.java

```diff
@@ -116,6 +124,13 @@ public DefaultTrackerBundleService(
         TrackerPreheat preheat = trackerPreheatService.preheat( preheatParams );
         trackerBundle.setPreheat( preheat );
 
+        Map<Enrollment, List<RuleEffect>> enrollmentRuleEffects = trackerProgramRuleService
+            .calculateEnrollmentRuleEffects( trackerBundle );
```

Comment:
In legacy importer, this was handled by async spring event. But this is synchronous and it will probably slow down the new tracker importer unless we make ProgramRuleEngine more performant.

Reply:
After a chat with Zubair I think we agreed that we are focusing on the functionality now.
Anyway, the loop to get the enrollment and event rules is done using parallel stream.

To really improve performance in the future we will need to refactor rule-engine to accept multiple entities that share the same context.

Reply:
Hi, if the code is no longer async, do we need to keep the `@TransactionalEventListener` annotation on `ProgramRuleEngineListener`?

Reply:
I think that when the new importer is working, we could completely remove the listener from the code.
In the legacy importer it should work as an async listener.